### DISCUSSION TITLE
tpc forwarding ssh connection not working

### DISCUSSION THREAD

**Discussion by phr34k** [rustdesk/rustdesk#8706] [General]
### Bug Description

I want use the tcp tunneling feature to forward an connection to the sshd(eamon). The problem when using ssh to connect the tunneled port, it is able to establish a connection and gather data for updating host fingerprints, but after printing the local version nothing happens anymore. 

Usually, after printing the local version it prints the remote version, followed by autentication. My best guess that something in the tunnel of rustdesk causes ssh to not read data from the tunnel, maybe settings related to window-length or the transmission queue. 

```
OpenSSH_for_Windows_8.6p1, LibreSSL 3.4.3
debug2: resolve_canonicalize: hostname 127.0.0.1 is address
debug1: Authenticator provider $SSH_SK_PROVIDER did not resolve; disabling
debug3: ssh_connect_direct: entering
debug1: Connecting to 127.0.0.1 [127.0.0.1] port 2222.
debug1: Connection established.
debug1: identity file C:\\Users\\phpfr/.ssh/id_rsa type 0
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_8.6
```

### How to Reproduce

In short, you'd need two machines, one machine configured with sshd and rustdesk, and the other machine configured with ssh and rustdesk. I'm using this from two windows-x64 machines.

the basic dependencies can be installed with:

```
winget install RustDesk.RustDesk
winget install Microsoft.OpenSSH.Beta
```

after that, setup rustdesk for unattended access on one machine and configure/install a vanilla ssh daemon, and use the other machine to connect to it. The github code suggests that `--port-forward` is intended to automate tcp tunnel forwarding from the command-line, however I've just done this from the user-interface. If you do it from the user-interface, localport: 2222, remotehost: 127.0.0.1, remoteport: 22. 

```
rustdesk.exe --port-forward XXYYZZWW 2222 127.0.0.1 22
ssh 127.0.0.1 -p 2222 -o IdentityFile=~/.ssh/id_rsa -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o User=server -tt -v
```

### Expected Behavior

The expected behavior is that the ssh is able to establish a connection and starts performing authentication and eventually opens up the tty and listens to stdin for interactive command prompts. When you connect ssh with the -vvv option, you'd usually very extensive debug output with atleast the remote protocol printed e.g.

```
debug1: identity file C:\\Users\\server/.ssh/id_dsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_9.5
debug1: Remote protocol version 2.0, remote software version OpenSSH_for_Windows_9.5
debug1: compat_banner: match: OpenSSH_for_Windows_9.5 pat OpenSSH* compat 0x04000000
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to localhost:22 as 'server'
debug3: record_hostkey: found key type ECDSA in file C:\\Users\\server/.ssh/known_hosts:1
debug3: load_hostkeys_file: loaded 1 keys from localhost
debug3: Failed to open file:C:/Users/server/.ssh/known_hosts2 error:2
debug1: load_hostkeys: fopen C:\\Users\\server/.ssh/known_hosts2: No such file or directory
debug3: Failed to open file:C:/ProgramData/ssh/ssh_known_hosts error:2
debug1: load_hostkeys: fopen __PROGRAMDATA__\\ssh/ssh_known_hosts: No such file or directory
debug3: Failed to open file:C:/ProgramData/ssh/ssh_known_hosts2 error:2
debug1: load_hostkeys: fopen __PROGRAMDATA__\\ssh/ssh_known_hosts2: No such file or directory
debug3: order_hostkeyalgs: prefer hostkeyalgs: ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp256
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
debug3: receive packet: type 20
debug1: SSH2_MSG_KEXINIT received
debug2: local client KEXINIT proposal
```

### Operating system(s) on local side and remote side

Windows 11 -> Windows 10

### RustDesk Version(s) on local side and remote side

1.2.6 -> 1.2.6

### Screenshots

see textual description of bug

### Additional Context

My eventual goal is to automate this and use proxy command of ssh to use rustdesk tcp tunnel as an intermediate connection to help administrator devices on constrained networks i.e. where latency is problematic.

~/.ssh/config
```
	Host xxyyzzww
		ProxyCommand powershell ~/rustdesk-ssh.ps1 -IdentityFile ~/.ssh/id_rsa -User server -LocalPort 2222 -RemotePort 22 -RemoteHost 127.0.0.1 -RustdeskId %h
```

~/rustdesk-ssh.ps1
```
param ($RustdeskId, $IdentityFile, $User, $LocalPort, $RemotePort, $RemoteHost)
write-host "${RustdeskId} through ssh on ${User}@localhost:${LocalPort} <---> $User@${RemoteHost}:${RemotePort}"
C:\Program` Files\RustDesk\rustdesk.exe --port-forward $RustdeskId $LocalPort $RemoteHost $RemotePort
ssh 127.0.0.1 -p $LocalPort -o IdentityFile=$IdentityFile -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o User=$User -tt -v```

**Comment by phr34k** [comment]
Please note that this is semi-related to https://github.com/rustdesk/rustdesk/discussions/2840, however this bug is specifically about the tunnel not working as intended. The old issue and discussion thread were about command line interace (not) supporting automatic connection and tunnel configuration like the present undocumented `--port-forward` and `--connect` aim at doing.

**Comment by phr34k** [comment]
Actually after a restart, I'm having better luck. I still have issues with tpc forwarding of rustdesk intermittently is very slow and gets stuck as-if it's not forwarding. It definitely happens intermittendly, and sometimes it helps starting a new console session or rebooting again.

There are also cases where it works straightaway, however as a cliffnote there's a problem with ProxyCommand in the OpenSSH package (https://github.com/PowerShell/Win32-OpenSSH/issues/2253) and the terminal effective is just frozen.

**Comment by phr34k** [comment]
After a lot of going back and forth, I have an working solution. I noticed after that my usage of proxy command might have been off, either just forwarding stdin and stdout to the tcp socket seems to work but requires a seperate program to be compiled. I haven't had any luck encoding the commands straight in powershell.

`ssh-forward.cs`

```ssh-forward.cs

using System.Net;
using System.IO;
using System.Diagnostics;
using Microsoft.VisualBasic;

namespace main {

    class Program
    {
        static void Main(String[] args)
        {
            Process x = new Process();
            x.StartInfo = new ProcessStartInfo();
            foreach (var arg in args.Skip(1))
            {
                x.StartInfo.ArgumentList.Add(arg);
            }

            //Console.Error.WriteLine(x.StartInfo.Arguments);
            x.StartInfo.FileName = args[0];
            x.StartInfo.UseShellExecute = false;
            x.StartInfo.RedirectStandardOutput = true;
            if (x.Start() == true)
            {
                int port = 0;
                String? line = x.StandardOutput.ReadLine();
                if (line != null)
                {
                    port = Convert.ToInt32(line);
                }


                var tcpClient = new System.Net.Sockets.TcpClient("localhost", port);
                tcpClient.NoDelay = true;
                var tcpStream = tcpClient.GetStream();
                var stdin = System.Console.OpenStandardInput();
                var stdout = System.Console.OpenStandardOutput();
                var stderr = System.Console.OpenStandardError();
                var tasks = new List<System.Threading.Tasks.Task>();
                System.Threading.Tasks.Task.WaitAll(stdin.CopyToAsync(tcpStream, 1), tcpStream.CopyToAsync(stdout, 1));

                x.CloseMainWindow();

                x.WaitForExit();
            }

        }
    }
}
```

`~/rustdesk-ssh.ps1` 

```
param ($RustdeskId, $IdentityFile, $User, $LocalPort, $RemotePort, $RemoteHost, $RemoteUser)
C:\Program` Files\RustDesk\rustdesk.exe --port-forward $RustdeskId $LocalPort $RemoteHost $RemotePort 2>&1>$null
Write-Host $LocalPort
```


`~/.ssh/config`

```
	Host XXYYZZWW
	    User server
	    HostName 127.0.0.1
	    Port 2222
	    StrictHostKeyChecking no
	    NumberOfPasswordPrompts 0
	    PasswordAuthentication no
	    BatchMode yes
	    IdentityFile ~/.ssh/id_rsa
		ProxyCommand ssh-forward.exe powershell -File ~/rustdesk-ssh.ps1 -LocalPort 2222 -RemotePort 22 -RemoteHost 127.0.0.1 -RustdeskId XXYYZZWW

```

**Comment by phr34k** [comment]
@rustdesk  I have a problem with the port forwarding of RustDesk. What I see in practice is that the tunnels seem unstable for an ssh session. Is there anything that could be done to improve the stablity? In general I don't really seem get much connection problems with rustdesk itself. 

Here's what happens: I have an port forward active, and connect ssh over the tunnel e.g. `ssh -vvv server@localhost -p 2222`.  About 10-20  into the session and it says the session was killed by the remote host. Note, I'm running ssh directly over the tunnel so in this context something must have happened to the tunnel.

I've tried to debug the ssh deamon, and in general I see that it's receiving and sending keep alive packets so the tunnel is actively used, and after some delay the client has disconnected the receives a keep alive timeout and closes the session on the server side. 

Another thing that I intermittantly happens is that it takes a very long time to establish the session over the tunnel. Verbose logging indicates it's waiting on remote authentication header, and it can take up to 30-40 seconds later until it's received.

I suspect, that both of these problems are likely related RustDesk. Is there anything that you could comment that could help explain the behavior that I'm seeing?

**Reply by MikeT-KBU** [reply]
Activate Auto Ping in SSH if possible all 10 Seconds. RustDesk kills the session on inactivity after 30? Seconds. Second is that you can only make one Connection over one Portforwarding in RustDesk. A Second connection can't or one of the two would be killed after seconds.

**Comment by kryztoval** [comment]
If the only problem is that the tunnel/relay may not be stable enough to keep an ssh session up then you may be better off  using something like mosh (The Mobile Shell) or another similar solution that does support reconnection to the session after a disconnection. Or, you could make your session inside a screen/tmux and reconnect to it everytime you log back so you recover the session upon disconnection.

It is not without limitations. Not sure if there is a windows version of mosh but pretty sure there are other solutions like that that do exist for windows.

**Comment by phr34k** [comment]
@kryztoval @MikeT-KBU Yes I've a tried lots of configurations with tcp keep alives, client keep alives and server keep alives most of it didn't had an real impact to keep the connection stable. At the time I was testing this, what seemed to be a factor is that rustdesk network was being upgraded to a new version. Just an fyi, I also came across problems with the remote desktop aspect just hanging and restarting on occasion. However, chrome remote desktop and teamviewer were working fine as backup's. It could be network thing on our end but it's just more likely it's a rustdesk (network-) specific in the sense that it's unstable.

In my situation, I just need a way to access machines for the rare but occasional software upgrades I provide. Most often these are scriptable i.e.  powershell-scripts and rehearsed. It's just easier to throw it against powershell than going through rdp-- so just rustdesk facilitating a stable network tunnel and maybe some command line versions of their software for tunnels is all that is nessecary, the rest is easilly done with sshd.
